[AST] timeline, divination, CDDT, overheal

otocephaly opened the pull request:
Updated the timeline to be more methodical and fixed minor arcana/neutral sect
Updated divination to be like lightspeed with some tweaks
Updated overheals to include delayed heals to show heals that are usually planned in advance
Consolidated CDDT to include some healing abilities and some dps abilities to show as a snapshot how the AST is managing their oGCDs
   - please note the weights and target are arbitrary so if you don't like them, lemme know and they can be adjusted fairly easily to what you think would be better
Updated changelog

For divination, since there aren't many logs that would contain this situation, I used this log to test overwriting divination
https://www.fflogs.com/reports/gGxcLpvwYC9FJVHf

Otherwise, it seems there is a bug with some of ASTs heals for exaltation, horoscope, and macrocosmos -- so while delayed heals seems to be functioning, it may look like it's not because fflogs hasn't attributed some of the heals of those sorts

otocephaly committed:
[AST] updating overheal for delayed heals

updating overheal section to include delayed heals for additional information since AST has three distinct moves that this could relate to

otocephaly committed:
[AST] consolidating cooldown list

consolidated cooldowns necessary for tracking since was seemingly separated with no value-add in each section

otocephaly committed:
[AST] reorganized action timeline to be more methodical

rearranged to be more methodical

otocephaly committed:
[AST] fixed actions and statuses

fixed actions and statuses to better fit action timeline and to have the appropriate cooldown for minor arcana
also updated neutral sect to be more descriptive

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
[AST] adding in divination section

adding in divination section similar to lightspeed, but more tailored to divination specifics

otocephaly committed:
[AST] updating changelog

updated change log with the listed changes

otocephaly committed:
Merge branch '6_0_CI_Divination_window_overheal' of https://github.com/otocephaly/xivanalysis into 6_0_CI_Divination_window_overheal

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+const BASE_GCDS_PER_WINDOW = 6
+const PLAYERS_BUFFED_TARGET = 8
+
+interface DIVINATION_Window {
```
style nit:
```suggestion
interface DivinationWindow {
```

otocephaly replied:
updated both this one and lightspeed with the same comment

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+				</Trans>,
+				severity: SEVERITY.MAJOR,
+				why: <Trans id="ast.divination.suggestion.usage.why">
+					About {this.otherAst.countOverwritten} <Plural value={this.otherAst.countOverwritten} one="cast" other="casts" /> of <DataLink action="DIVINATION" /> were overwritten resulting in a loss of at least {this.parser.formatDuration(this.otherAst.timeOverwritten)}.
```
It would be better to remove the "About" here.  Even if there is a little bit of fuzziness with respect to possible overwrites, we should tune the sensitivity for counting to err on the side of not counting it as an overwrite unless we're sure so that we can be confident with the number of overwrites we're reporting here.

otocephaly replied:
removed about
the only doubt I had was the upper limit, not the lower so I thought it was conservative anyway

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	private tryCloseWindow(event: Events['statusRemove']) {
+		//note: closing other AST window handled in open window since closing the window doesn't catch what we need because status remove happens before status apply
+
+		//to stop using function if window isn't even open. used when splicing
+		if (this.currentWindow == null) {
+			return
+		}
+
+		//close window and push to history
+		if (event.source === this.parser.actor.id) {
```
I think it would be better to limit this function to checking if we need to close the window here based on the source, and then passing the timestamp to a second function that actually does close the window.  Then the "close when your window gets overwritten" handler up in tryOpenWindow could just pass the timestamp to the actual close function as well and you don't have to synth a full event for it, while being a little more clear about why you're only trying to close certain windows.  (And similarly onComplete wouldn't have to synth an event either.  You could also just use this.parser.currentTimestamp in the actual close function and never pass around the event.timestamp).

otocephaly replied:
added function to do as described and updated the relating splices as a result

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	// just output, no suggestions for now.
```
Comment is incorrect, as we are generating suggestions.

otocephaly replied:
removed oop

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+		//for now, the message is copied directly from DivinationDowntime
+		const message = <p><Trans id="ast.ogcd-downtime.divination.description">
+			<DataLink action="DIVINATION" /> provides Astrologian with a strong amount of raid DPS when stacked together with arcanum.
+			Damage percentage bonuses stack multiplicatively, so it's most optimal to stack it with cards from <DataLink action="DRAW" /> and an Astrologian's <DataLink action="ASTRODYNE" /> when 3 unique seals are obtained. <br/>
+			Try to time the usage to match raid buffs and high output phases of other party members - it's more important to use it on time rather than hold it.
```
Somewhat contradictory copy here, since on the one hand you're saying "use Divination when you have draw cards out and 3 Astrodyne seals" and on the other hand you're saying "use Divination with other raid buffs and party burst windows".  Maybe just remove the line about "damage percentage bonuses stack multiplicatively..."?

otocephaly replied:
Originally with the wording I was hoping it would come across as aim to use arcana, astrodyne, and divination along with party burst windows, but I see that it did not come across as such
Removed line as suggested for clarity

AzariahAeroborn commented on src/parser/jobs/ast/modules/Divination.tsx:
```diff
@@ -0,0 +1,297 @@
+// the character selected for analysis. windows that clip into
+// AST Divination will be marked.
+// Used DNC Technicalities as basis for this module. Rewritten from previous module for consistency purposes
+export default class Divination extends Analyser {
```
No default export.
```suggestion
export class Divination extends Analyser {
```

otocephaly replied:
I get this error when I do that
![image](https://user-images.githubusercontent.com/89435408/145928480-c4737c5f-e48b-414d-8b6d-8fd843cc0c8d.png)

otocephaly replied:
pls ignore this -- resolved in other conversation

AzariahAeroborn commented on src/parser/jobs/ast/modules/oGCDs.tsx:
```diff
@@ -0,0 +1,125 @@
+	DPS: 'dps',
+}
+
+export default class oGCDs extends CooldownDowntime {
```
No default export
```suggestion
export class oGCDs extends CooldownDowntime {
```

otocephaly replied:
I get this error when I do that
![image](https://user-images.githubusercontent.com/89435408/145928455-dac5cda4-83aa-4ddb-a20f-aca9af3c1ff2.png)

AzariahAeroborn replied:
You will also need to update the class index to do a named import (import {oGCDs} from ...) instead of a default import without the braces.

otocephaly replied:
works beautifully
thanks!

AzariahAeroborn commented on src/parser/jobs/ast/modules/oGCDs.tsx:
```diff
@@ -0,0 +1,125 @@
+	override trackedCds = this.COOLDOWNS_TRACKED
+
+	override addJobSuggestions() {
+		this.COOLDOWNS_TRACKED.forEach(cooldownGroup => {
```
I'm not a huge fan of this approach, since it can potentially generate 6 different suggestions if the user is missing targets on all of the skills separately.  Generating multiple suggestions implies to the user that using cooldowns is 6 times more important than something else that generates a single suggestion, and especially to the extent that we usually allow for a lot of not-use on healing cooldowns it seems like a poor signal to send.

I would also prefer to see the suggestion calculations be based on percentUsed (invert the thresholds, basically), in order to match the way the checklist threshold is handled, so that the calculations are the same in both halves of the module, for future maintainer ease of mind.

otocephaly replied:
Updated to consolidate healing suggestions since subjective anyway
Thresholds updated to match

Updated suggestion calculation using percentUsed

AzariahAeroborn commented on src/parser/jobs/ast/modules/oGCDs.tsx:
```diff
@@ -0,0 +1,125 @@
+	@dependency private suggestions!: Suggestions
+
+	override checklistName =  <Trans id="ast.ogcds.name">Use your OGDs</Trans>
+	override checklistDescription = <Trans id="ast.ogcds.description">Consider using the below actions to help support your party with damage buffs and reduce the amount of GCD healing necessary. </Trans>
```
"Consider using..." is pretty squishy.  Can this copy be improved to be more assertive?

otocephaly replied:
Updated along with healing suggestions

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
[AST] default clean up

cleaned up based on feedback

otocephaly committed:
[AST] updated Divination and Lightspeed based on feedback

- removed default
- updated variable name
- consolidated actual closing of window in one function rather than splicing

otocephaly committed:
[AST] updated cooldowns

separated how dps and healing actions are handled
reworked thresholds a bit to reflect
removed unused content as a result

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
[AST] yarn extract plus copy

used yarn extract and copied the translations for the .copy" items

otocephaly committed:
[AST] updated oGCDs checklist

- Added oGCD_HEALS and oGCD_DPS sections as part of the checklist
- directly copied CDDT to be able to split dps and heals since using same module overwrote the other

otocephaly commented:
Please see draft checklist with split healing / dps

![image](https://user-images.githubusercontent.com/89435408/147020400-ef4f8acd-4d8f-43f7-904a-aae7cb224250.png)

otocephaly commented:
I have a few more changes coming because someone pointed out that Macrocosmos is actually GCD not oGCD oop

otocephaly committed:
[AST] oop macro no oGCD

updated macrocosmos to be gcd not ogcd as mistaken

otocephaly committed:
[AST] harmony of body addition

added harmony of body to cast time due to speedmodifier since I was messing around in actions anyway

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
[AST] divination less punishing

added allowed average downtime because it would otherwise be too punishing
it was removed before because I had it lumped with macro and I thought it would be better to have theoretical max, but now that it's on its own, I think it's better to include it

otocephaly committed:
[AST] update changelog to reflect missed changes

updated based on macrocosmos and oGCD heals

otocephaly commented:
should be good to review again

suShirow commented on src/parser/jobs/ast/modules/CastTime.tsx:
```diff
@@ -0,0 +1,28 @@
+	override initialise() {
+		super.initialise()
+
+		const presenceOfMindFilter = filter<Event>()
```
var naming nitpick, can we modify this filter from WHM to reflect the AST's Harmony of ___ by name?

otocephaly replied:
lol not nitpicky at all
I should use my eyes when copying instead of committing and praying lol
it's been updated

otocephaly committed:
[AST] fix macrocosmos in timeline

macrocosmos was doing something weird so I removed it, but then it did weirder things today so I keep it

otocephaly committed:
[AST] updated based on feedback

updated based on feedback

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

AzariahAeroborn commented on src/parser/core/modules/CooldownDowntime_copy.tsx:
```diff
@@ -0,0 +1,350 @@
+//direct copy of CooldownDowntime (CDDT) to allow for a second use specific to secondary role actions (healing/dps, etc.)
```
#1451 added a way to do these defensive/healing only role actions, can you either update this PR to use those changes to core CDDT or separate this change into a future PR that uses the updated core CDDT?

AzariahAeroborn replied:
@otocephaly No, sorry, I'm asking to remove this copy and adapt your CDDT usage in light of the updates to core.

otocephaly replied:
done! at least it was easy to do
thanks for that!

otocephaly committed:
Merge branch 'endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
match copy with recent commit

match copy with recent commit

otocephaly committed:
add copy to end of class name as required

jokes needed this to work too lol

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into 6_0_CI_Divination_window_overheal

otocephaly committed:
[AST] redoing structure for OGCD heals

updated structure to not copy CDDT and instead use defensive portion of recently-added code

otocephaly committed:
yarn extract

yarn extract
